router/usecase/pools/routable_cw_orderbook_pool.go

```diff
@@ -0,0 +1,246 @@
+	poolmanagertypes "github.com/osmosis-labs/osmosis/v25/x/poolmanager/types"
+)
+
+var _ sqsdomain.RoutablePool = &routableOrderbookPoolImpl{}
+
+type routableOrderbookPoolImpl struct {
+	ChainPool     *cwpoolmodel.CosmWasmPool   "json:\"pool\""
+	Balances      sdk.Coins                   "json:\"balances\""
+	TokenOutDenom string                      "json:\"token_out_denom\""
+	TakerFee      osmomath.Dec                "json:\"taker_fee\""
+	SpreadFactor  osmomath.Dec                "json:\"spread_factor\""
+	OrderbookData *cosmwasmpool.OrderbookData "json:\"orderbook_data\""
+}
+
+// GetId implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) GetId() uint64 {
+	return r.ChainPool.PoolId
+}
+
+// GetPoolDenoms implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) GetPoolDenoms() []string {
+	return r.Balances.Denoms()
+}
+
+// GetType implements domain.RoutablePool.
+func (*routableOrderbookPoolImpl) GetType() poolmanagertypes.PoolType {
+	return poolmanagertypes.CosmWasm
+}
+
+// GetSpreadFactor implements domain.RoutablePool.
+func (r *routableOrderbookPoolImpl) GetSpreadFactor() math.LegacyDec {
+	return r.SpreadFactor
+}
+
+// CalculateTokenOutByTokenIn implements sqsdomain.RoutablePool.
+// It calculates the amount of token out given the amount of token in for a orderbook pool.
+// Fails if:
+// - the underlying chain pool set on the routable pool is not of cosmwasm type
+// - token in and token out denoms are the same
+// - the provided denom pair is not supported by the orderbook
+// - fails to retrieve the tick model for the pool
+// - runs out of ticks during swap (token in is too high for liquidity in the pool)
+// - `TickToPrice` calculation fails
+func (r *routableOrderbookPoolImpl) CalculateTokenOutByTokenIn(ctx context.Context, tokenIn sdk.Coin) (sdk.Coin, error) {
+	poolType := r.GetType()
+
+	// Esnure that the pool is a cosmwasm pool
+	if poolType != poolmanagertypes.CosmWasm {
+		return sdk.Coin{}, domain.InvalidPoolTypeError{PoolType: int32(poolType)}
+	}
+
+	// Get the expected order directionIn
+	directionIn, err := r.GetDirection(tokenIn.Denom, r.TokenOutDenom)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	directionOut := directionIn.Opposite()
+	iterationStep, err := directionOut.IterationStep()
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	// Get starting tick index for the "out" side of the orderbook
+	// Since the order will get the liquidity out from that side
+	tickIdx, err := r.GetStartTickIndex(directionOut)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	amountOutTotal := osmomath.ZeroBigDec()
+	amountInRemaining := osmomath.BigDecFromSDKInt(tokenIn.Amount)
+
+	// ASSUMPTION: Ticks are ordered
+	for amountInRemaining.GT(zeroBigDec) {
+		// Order has run out of ticks to iterate
+		if tickIdx >= len(r.OrderbookData.Ticks) || tickIdx < 0 {
+			return sdk.Coin{}, domain.OrderbookNotEnoughLiquidityToCompleteSwapError{PoolId: r.GetId(), AmountIn: tokenIn}
+		}
```

Comment:
I'm worried about performance. Imagine that many users keep playing with the router, trying to estimate a large amount in that is guaranteed to exceed available liquidity in the direction of the swap.

With the current design, every time users type something different, we would iterate over all available ticks before reaching the end and concluding "no available liquidity".

Is it possible to store an accumulator that aggregates available liquidity in each direction? That way, we could use it to understand if the given token in amount exceeds the available liquidity immediately without having to iterate until the end.

Please let me know what you think.

Reply:
I think we can precompute amount in required to consume liquidity on all of the ticks and we can check on that. That has to be done from the ingester side then.

Reply:
With that we can really index each tick by it's accumulated required amount in and binary search the tick to compute the last tick filled and avoid O(n) from router all together. Not sure if that's an overoptimization or not though. But I think it's possible.

Reply:
@iboss-ptk I agree that the second approach is more performant but unlikely to be launch blocking. 

That being said, I think it would be beneficial to have this algorithm created so that we can minimize quote tail latencies. Performance of the router is critical.

Estimates over CL pools had been observed to be the bottleneck due to similar iterations, and I think efficiently designing orderbook will pay off down the road while also giving the flexibility to transfer this architecture to CL.

Let's break this down into 2 stages:
- Stage 1: only preventing not enough liquidity cases from iterating through all the ticks
- Stage 2: the aggregation and binary search mechanism you proposed

Please let me know if that sounds good.